### ISSUE TITLE
Modulefile (tcl, lmod) generation fails, '... Cannot create a consistent method resolution...'

### ISSUE THREAD

**Issue by hartzell** [spack/spack#7452]
Trying to generate Lmod modulefiles fails, see below.

### Expected Result

Lmod module files for installed applications.

### Actual Result

```
$ spack find
==> 35 installed packages.
-- linux-ubuntu14.04-x86_64 / gcc@4.8 ---------------------------
autoconf@2.69    expat@2.2.2       libssh2@1.8.0            m4@1.4.18       readline@7.0
automake@1.15.1  gdbm@1.14.1       libtool@2.4.6            ncurses@6.0     sqlite@3.22.0
bzip2@1.0.6      gettext@0.19.8.1  libxml2@2.9.4            openssl@1.0.2n  tar@1.29
cask@0.8.1       git@2.15.1        lmod@7.7.13              pcre@8.41       tcl@8.6.6
cmake@3.10.2     libbsd@0.8.6      lua@5.3.4                perl@5.24.1     tree@1.7.0
curl@7.56.0      libiconv@1.15     lua-luafilesystem@1_6_3  pkgconf@1.4.0   xz@5.2.3
emacs@25.3       libsigsegv@2.11   lua-luaposix@33.4.0      python@2.7.14   zlib@1.2.11
$ spack module refresh --module-type lmod --delete-tree -y
==> Regenerating lmod module files
==> Warning: Could not write module file [/home/hartzell/spack/share/spack/lmod/linux-ubuntu14.04-x86_64/gcc/4.8/autoconf/2.69.lua]
==> Warning: 	--> Cannot create a consistent method resolution
order (MRO) for bases SpecBuildInterface, Spec <--
==> Warning: Could not write module file [/home/hartzell/spack/share/spack/lmod/linux-ubuntu14.04-x86_64/gcc/4.8/automake/1.15.1.lua]
==> Warning: 	--> Cannot create a consistent method resolution
order (MRO) for bases SpecBuildInterface, Spec <--
==> Warning: Could not write module file [/home/hartzell/spack/share/spack/lmod/linux-ubuntu14.04-x86_64/gcc/4.8/git/2.15.1.lua]
==> Warning: 	--> Cannot create a consistent method resolution
order (MRO) for bases SpecBuildInterface, Spec <--
$
```

### Steps to reproduce the issue

See above.

### Information on your system

This includes:

 1. which platform you are using

```
$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.1 LTS"
```

 2. any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.)

I'm on the `develop` branch at commit: b91b00f6f954e6d4200ff38e845b2dc2cb256415.

I'm using a config based on an older environment, there's some risk that I'm not up to date....

`etc/spack/modules.yaml`:

```yaml
modules:
  enable::
    - lmod
  lmod:
    core_compilers:
      - 'gcc@4.8.2'
    hash_length: 0
    whitelist:
      - gcc
    blacklist:
      - '%gcc@4.8.2'
    verbose_autoload: false
    all:
      filter:
        environment_blacklist: ['CPATH', 'LIBRARY_PATH', 'LD_LIBRARY_PATH']
      environment:
        set:
          '${PACKAGE}_ROOT': '${PREFIX}'
    ^python:
      autoload:  'direct'
      filter:
        environment_whitelist: ['LD_LIBRARY_PATH']
```

**Comment by hartzell** [comment]
I thought the problem might stem from the fact that this system has 

```
$ /usr/bin/python --version
Python 2.7.6
```

but using spack to build `2.7.4` doesn't seem to fix the problem.

**Comment by scheibelp** [comment]
FWIW as a data point:

I do not get this error with `spack module refresh --module-type lmod --delete-tree -y` when copying your `modules.yaml` (except modifying the core compiler) running on latest develop 82aa54151a

I'm running with python 2.7.10 on a RHEL-type system. This commit is equivalent to what you were trying with so I don't think it's a matter of code difference.

**Comment by hartzell** [comment]
Am I correct in assuming that prepending a spack built python to my path is sufficient for it to be used when I run the spack tools?  The spack command itself starts off with `#!/usr/bin/env python` so I'm 99.99% certain that this:

```
$ which python
/home/hartzell/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/python-2.7.14-qyytendyiuypwizustouwidjta4nhoce/bin/python
$
```

is sufficient to rule out my system's old python.

**Comment by hartzell** [comment]
@scheibelp -- have you installed the same applications that are giving me heartburn (e.g. `git`)?

**Comment by scheibelp** [comment]
> scheibelp -- have you installed the same applications that are giving me heartburn (e.g. git)?

I have not. Sorry I won't be able to look into this again for several days and realize https://github.com/spack/spack/issues/7452#issuecomment-371993861 isn't particularly helpful but I'm in a rush

**Comment by hartzell** [comment]
No worries, I appreciate the peek that you did take.

Spack generates modulefiles for *most* of the things that I have installed, those three (autoconf, automake, git) are the only ones that seem to cause a problem.  git is the only one I itch for.

**Comment by hartzell** [comment]
A bit more data.  This problem occurs when regenerating the *tcl* modulefiles w/out any custom `modules.yaml` file.

Oddly enough, installing `git` generates a modulefile for it:
```
$ ls share/spack/modules/linux-ubuntu14.04-x86_64/
[...]
git-2.15.1-gcc-4.8-5cjlsjd	  ncurses-6.0-gcc-4.8-ddmfsab		   tree-1.7.0-gcc-4.8-pu3obj5
[...]
```

But regenerating the modulefiles throws an warning and the modulefile is missing:

```
$ spack module refresh --module-type tcl --delete-tree -y
==> Regenerating tcl module files
==> Warning: Could not write module file [/home/hartzell/spack/share/spack/modules/linux-ubuntu14.04-x86_64/autoconf-2.69-gcc-4.8-o5o33x2]
==> Warning: 	--> Cannot create a consistent method resolution
order (MRO) for bases SpecBuildInterface, Spec <--
==> Warning: Could not write module file [/home/hartzell/spack/share/spack/modules/linux-ubuntu14.04-x86_64/automake-1.15.1-gcc-4.8-ihip7gg]
==> Warning: 	--> Cannot create a consistent method resolution
order (MRO) for bases SpecBuildInterface, Spec <--
==> Warning: Could not write module file [/home/hartzell/spack/share/spack/modules/linux-ubuntu14.04-x86_64/git-2.15.1-gcc-4.8-5cjlsjd]
==> Warning: 	--> Cannot create a consistent method resolution
order (MRO) for bases SpecBuildInterface, Spec <--
$ ls share/spack/modules/linux-ubuntu14.04-x86_64/
[...]
gdbm-1.14.1-gcc-4.8-44cevti	  m4-1.4.18-gcc-4.8-r26rlmg		   the-silver-searcher-2.1.0-gcc-4.8-tsrpdma
gettext-0.19.8.1-gcc-4.8-yvklhmb  ncurses-6.0-gcc-4.8-ddmfsab		   tree-1.7.0-gcc-4.8-pu3obj5
libbsd-0.8.6-gcc-4.8-5pyiwos	  openssl-1.0.2n-gcc-4.8-v2gvwhg	   xz-5.2.3-gcc-4.8-ekob74s
[...]
```

**Comment by hartzell** [comment]
One more Sunday comment: if the the `modules.yaml` file included [in the initial problem descriptions](https://github.com/spack/spack/issues/7452#issue-304025785), installing `git` sucessfully generates an **Lmod** modulefile.

The problem seems to be isolated to regeneration.

**Comment by hartzell** [comment]
@alalazo -- does this ring any bells?

**Comment by alalazo** [comment]
@hartzell This should be solved by #7457, can you confim it? It's a bug in `SpecBuildInterface` that is revealed by module generation.

**Comment by hartzell** [comment]
I can now generate tcl and lua modulefiles.  **THANKS!**
